project/api.py

```diff
@@ -0,0 +1,204 @@
+from datetime import datetime, timedelta
+import flask
+from flask import request, jsonify
+from werkzeug.exceptions import HTTPException
+import pandas as pd
+
+
+app = flask.Flask(__name__)
+
+
+def to_milliseconds(dt):
+    """Convert datetime object to timestamp in millisecond format"""
+    return int(datetime.timestamp(dt)*1000.)
+
+
+def prepare_request(time_range_start, time_range_end):
+    """Convert datetime time range to  timestamp in millisecond format"""
+    datetime_start = datetime.strptime(time_range_start, '%Y-%m-%d %H:%M:%S')
+    datetime_end = datetime.strptime(time_range_end, '%Y-%m-%d %H:%M:%S')
+    datetime_start_code = to_milliseconds(datetime_start)
+    datetime_end_code = to_milliseconds(datetime_end)
+    return datetime_start_code, datetime_end_code
+
+
+@app.route('/', methods=['GET'])
+def home():
+    return "<h1>Junior Data Scientist/Data Engineer Test Task</h1>"
+
+
+@app.errorhandler(Exception)
+def handle_error(e):
+    """Generic exception handler"""
+    code = 500
+    if isinstance(e, HTTPException):
+        code = e.code
+    return jsonify(error=str(e)), code
+
+
+@app.errorhandler(404)
+def page_not_found(e):
+    """Handle cases with empty query parameter"""
+    code = 404
+    return jsonify(error=str(e), code=code)
+
+
+@app.errorhandler(400)
+def handle_bad_request(e):
+    """Handle cases with wrong query parameter"""
+    code = 400
+    return jsonify(error=str(e), code=code), code
+
+
+@app.route('/api/v1/queries/avg_elapsed_time', methods=['GET'])
+def avg_elapsed_time():
+    """Returns an average elapsed time for all queries on certain date"""
+    query_parameters = request.args
+    date = query_parameters.get('date')
+    if not date:
+        return page_not_found('No date provided')
+
+    data = pd.read_csv('queries.tsv', sep='\t')
+    try:
+        date_start = datetime.strptime(date, '%Y-%m-%d')
+    except ValueError as e:
+        return handle_bad_request(e)
+    date_end = date_start + timedelta(days=1, milliseconds=-1)
+    date_start_code = to_milliseconds(date_start)
+    date_end_code = to_milliseconds(date_end)
+
+    day_data = data[data['time'].between(date_start_code, date_end_code)]
+    if day_data.shape[0] == 0:
+        return handle_bad_request('Date is out of range')
+
+    day_pivot = day_data.pivot(index='query_id', columns='status', values='time')
+    not_closed_within_day = day_pivot.isnull().values.any(1).nonzero()[0]  # if not empty returns indexes
+
+    if not_closed_within_day.shape[0] > 0:
+        day_pivot.drop(day_pivot.index[not_closed_within_day], inplace=True)
+
+    day_pivot['elapsed_time'] = (day_pivot[1] - day_pivot[0]) / 1000
```

Comment:
It is always better for clarity and readability to use column names instead of indices.